### DISCUSSION TITLE
Show previously rendered PDF instead of showing 'Loading PDF...'

### DISCUSSION THREAD

**Discussion by contentkonker** [wojtekmaj/react-pdf#1055] [Q&A] [answered]
### Before you start - checklist

- [X] I followed instructions in documentation written for my React-PDF version
- [X] I have checked if this bug is not already reported
- [X] I have checked if an issue is not listed in [Known issues](https://github.com/wojtekmaj/react-pdf/wiki/Known-issues)
- [X] If I have a problem with PDF rendering, I checked if my PDF renders properly in [PDF.js demo](https://mozilla.github.io/pdf.js/web/viewer.html)

### Description

Hi. 

I built an exact copy of the [repl page](https://react-pdf.org/repl?code=3187b0760ce02e00408a057025803c450298c0bc300500943807cf805030c00f0022230080b6198531165d400a02180e659a004f003619b006f61622003a0046200099080beed3a72a005430a58d3c54a8a231c8016187928c009d54c0066a83128d9b3803f1801856d5cc25180077242873187f1e6028005a00072547184f0e4d2a007a5d7d7774ec8313314943392830b1753a307864002b742c1b18111e1800591e3060731e4cfcdcad7c984362c2b3591e0470907b6236a43e040c1118269f3b00374ecc083ebd3634ad0049467e0c43cd11e35339245381550bce085b606c00220cbbb3880c80473a834008cb25abc4f8ef2786406941d3ed8663518dd6410043c8ca500a8c3383e1e3c400b75004088403023800b86a58282d87850569345a305018094486012000cfd5210c030b53b07239a49b0ad1c3c4608020a4f3120943c111f149345913d2888243d530ab2c1323a5d1e93cf6392350cae2559261f4ea354c000a2d50435444083e0f16c3ae6ab5f5dd1e000693d0c21292c020463c9fc3050cc001483b293a233796d99801e8e1a3c9c46021baa4b286118f152402b01b2406df1ad274c165f2c548b7552cf4b4c0002f568b0603c6812044625b08003d6775bb6ddb05821fd917a3e3aa8e16a8060bcb77d799337f12866b21800154c0ad101f619580552a608c0021c40605b602b44bb6bbed8c0f180c4015f176772bdd7cbf5f43dd198019284b092d7af225846db1366070c17bc8d614a7eada622b0fcb76cd136659d82f84043a2a0017f54f102a61920d19343c0008fd61ca082d29875e9ba30e452a7863ee02a28c256ce8613499830091b60ec9eaa235b4471ab23c36e768acfe3400d91e6b9742e8f1c30f0c1ada4d06ccb3ba9800677888ed931a5b2c2ea6e827baad0b4d7b38663d2d78b2c3020d7891ac020f39368c880b68b11024a01ac14a256265f2ac246100169d2b4914d6e02d208729ce4de664cc48336a49feb6b38a1b6eba18095985c95deef8c04eab412bde4b259efa25de63993bf8ac14ebcb95c308011b913c2da25946a49f6181ba8975958460e47e94b995d524aad836cc84a45a4df10b4ed806bd776f8ada2c040bd8b48972cd3b00b39ae928c0501863c2254c916f28d8b4b6e0018ad8482d5d6476b767a6012cd000eb6ad6a7a924e6c11c9744b3355d951485f20c534a0216620a05cb54fe07253a458bac3eeae1c8666940b0b514c7340e3f48000370c07c2d2d3758b62ba8847abc84a0b1cd4c9deedb6e1a9d89d8d81c87a0fae3674b08541dd28462c18662e7a1aad4c38c0a4a20662cb782f59dbd9328695da2ae5569482385cd7448076ce9369339df24c6d54d8d01602569266300cf7c82c0104393abd516b60716000a44ad2d78e9c4fe22c4c8c3234e8585982b27605b822ea2d29c3526ad1793f13c67c829b69dc1efd2827d009946df4d280d83e779308740555609d932fbad8b4872df9280025cacd1ac5efa2add40bd5bc44eb9dcdab401edd46964fb36270be808a98488c896bec3680b769e18277dc0608975d86b755f66081d46eecb98329ca00e054dad2e775e2ac57086daf412052a92caf6f54e31e795e3a8acd4480442c5f2b0dcd06068e220900462f236d2e82a7ec214bd99d6d81559c0371e414c5cb9d2de5810f1962509c9e68ed4f4e742caf756076c563f2414ec9ef8ea142ad0d98b52ecd58993581626005e8d20928dd040207121c8583ca726bc8069f06c1ad1e9110e728f9e8bdf1da1f57915b3a4e44edb5435668252ab22e4e78730f46bc3d05f1ba50af98ebb1e5b4915c32d2102b48031096a1619e22d20f4c03406853bc08406ad810a4557a9883b07c2bb0407c4cd0c406c4e46398ab76611ca09b030682ae4100591e84ec3d28a302de29a3e0a6ccc2c022c50cd78c42da2121808ea74581d51e424913afd49b2f5386bdda51f23141281f8c007acb1c867439a001957b16c724b9266078ea12d19e87d2686246c54a5a13c04b11e6bce359e97461cd79692cd5c2ce48437640ae70059e4829d509a2a75b0365bc5fe664939593f53104342a800e8e13340ac17f0b1830038a5c8f8f4861268b231c3e241cab1c01344ecd997d2b93d20247a11e7da1001d9841009489922cf6154386391385d51d6b9e74ac8b439708c06bd604ac5382f8188b14883c15b37e160f386e7e4ec53d5aa9390c5cc063bd401e3790b8313d6298a53d15a1f49c312026ed512269c801959ba40b5820a82fb3238cb057ab695650c9cb39163e410d91967dccc0d09251b2e9df894ca315811c0cc3280833a0065e44d87a3007f95ccad66e186075980a33b8925fbaf547554abd2da0d8f2a299531b0ecc47bf4534f09cd18c528f3db12501f0e00754615e4c63d09a105924bc00093d9c143b0ac22c92c9c640b4e07409b90ef9420b49981d2221e0b762c01ec238f0bb0364cd4ea48ae15a8bba10e22c581944dcca2dba36231d10ac9a674adab3b4f0b8c0c41d4bd53a001059eb404e4980641be745660731842b2964188404e4248b69f6100db8001c9db6bead1b32f21299d8818614b083230aca98d000859a67417aab02073c8164ebe42560b1bd3aaac56f87c0baa257763317d6742b6e78783d962631c004d63447199b788d31df228ba1571c7905812da9d3849a4d549786181d845a03430e3e7647499e95af02a1c688c814e69c9c911854103f6b54312a6dc0e52cfeb9a3d2a21422020716ad68a33982954a2932991755cdf20c41761d2ee5cc64d0059a252ad18015ebe45368c0735cea9c3b16265d29c46dc9814ce54d6824e3332b10ac2ca66b1109b405ac5713181303789da3b2624552d6fb9308d6721761c67560eee3aac628fddcbb6f9175fc1a4660095e42c8ca17d26ca9d201334ed693492c51243aa660212a67941006c4ca1a3bb17aaa58e5ac7b9960a8c9281d060f6b928ba3741e89f95706287c4c6347a0fdd29b5849c9b8b6b540e97c0000fcb076c2edda8e6840178561c4c2a52332bf2703b9a5c77cde02211d71d1b0b688e0649689800c5cb180d3b4e30550a001c5260b4726a70e2d982c087bd921e2bc0190afc4c5268bfc8f94d94b1e4f7c9e68557aba4cc96730687e854f359c7f61fb4d314236deab834f13a7fbb8d72f56e50854aa5595b001982cfcea8b765a0cc00cb7bf15cd0bbd61f10ddbbb5383dbf35c691d0ab9eaec70d942cb1646c91066eb099b9c9983bb60c436fc3df2cd359313743e94447f8998df939f254a7ad391f6f636ccab62a1280855430a12c02028d8c076c6e955266634071cdd9a444d716795a2800bd7183a4886610f1aeee40185ba62f257ed80360911fdd3b189310310b9728ac7354206f1080b6800a586c9180840f0b9ac0006428e6469f1b505371293e8894f4fbedf3e368da38d71a8a5ce18c660b10661fdd46e62b904f31429f46d831ad7350138bf1513e67998d57e774dd1c0b720614db8000caa14e491414d32714cd57c761d63d278b35531095846ebf7b1c4c872f8cf4756e04b12cbcbc95eecbad06cd64c0042f48a6697ccc1f2be72f5bdc7d97bdce408f2b7d76494f240319b0d50e5269b20804b11d84ce42c29410125d49a2ad56049f65830e5c573c70922a1937b859324678322721e6b05b47c10aa69333d0bc58c7f32935b90c26941a27946993002536146590f10c05fd7834b15b42f774c4ba42a7402eb16134a7c0c7271d1ce5563c52744dc3370f00b771e6b769e5b77185447447280c00703161e00006b48d5f03c44246245242380a42a433a73e6c36ba67a47518a05463e7836194e17d14e0d9a5a94ce0c916049165418d69606840d31e2341383382311b70b4203790a315780f82f87b83307f84044c0000265907880c94844cc68460d49e50d6447b74774cc000097c94d10b155e58d5727f06a0fbe6c63ea0b617a04c45877464c37e6bc7cc500aec5ce02575f0c03d67ee1196da4cb22121c1b17730bc5390534cd3cb19f0954ae1802f774b57a18e3fa68861648b7512153127a45b079436c34641a22268642971501b32c0a96aa71888152d0804b54de33a2586125d8aae28e7f076c46f2f345954d4be6de0d52a8c667901720b88977a85b17ff0c01f3478f5e3280807142b55b06dc0004d1f17c465d1d97744dd6d43133b06947f33477dc52d64a71a3ab340f8d7227837dc3645d452c6436c529a33d56d7dc21c24d5af0cd3412da0627470ca4720547a207e1323c5012268000296247be37c13c14c6c11492e76bc5446bc77447b1bf0f0190b5151648c000253be7dc3421c8fb218a4e15381582a942e6a15d27944271311256c141301661c5b03746ff30c434ce116c76c25a7036ba3389ce7141104932c06ac50003115830b2110d485a0c74cc27201b5643a73b840670048e7ee46611045001c1b0ed268e2720605e8c748926010fa25b2ee32f4424125990e44fc7ce312378ee667b03d7680c23c4160520c24f4f46b5f4e394837ea0d95c65091102806c11ee76883f51113b47631046d2bf6511ad22c45c0535946e17b51962330e20d576c7943e442c7921e06f14706aa3a09bc1d337a73c02c6e0dc49170aa14c44217938757c49a12c03270043514c13c00119170123cac321d5ce0cd28c74a1b01fd30943331c7dceab6a8707758b1778c6f15ee3567833cb65ce0b0a7360830809584b148ab0167934e812a33cd3a50c72d5873cd01c684b9cee8d2a010f4e0606ed7120ae4609982314e1330302db01b0a643747dc48a1920525720ec9751822c051cf1d18d0c43442b00f420f0405c901c78849f28507b0c25ee7c4a1523960f7a02e57830e0d680c5144c67412f78ca1016638db058a5828f47fc91372f1b13b834e5ea73a12248a21c1806757a43e72680e0a945f20d10b8c8480b762f8c3005e4be0758af774ce150b26ae182613472abf7d71c31e21920d746432cafc8bf16c242c12768aba2ce4367d93792c02ef1531606c0870e4f6683a646f0f1524302e6d6a8f836ac314a5532649a272602d8a07956ca040000e5b45e45191da458694ce13bc2616f92f37f10c02d80f45b2f835e378331231765f3b05fd47c7f024f2151566a85e33cb1641da68042a19f3389b488ad14a535f6382d10d1e1342b710d1f0b0d39e6b4550b04e4f654a3f2664548b22fa58302af8e6d813c6dafa263cc37712042a1525ae660973fb33320b5394a440014020007e8d47e55581f3684f1aa40d1a6613ba26c5a8a2c7620164150b20ee3747a897456c2f414a258eb022759405267a6c33f22707a96942d12c8a861561107ed7cf102a6c7b647671d2c068c22c0c929c746c874582bce4b584954b45012c32add530d9cb972685381a2724a6d1ce34a9e00007e4a469d74b1bd482bc35a61290605395b4556c3c3cc31d0213b4d687145c688aaec0345685e433012261452a2b0b7a2621987993fa37f14ae726a68e52a6d8754a65e466356ec934b0c012c4863a34fc66a4dc18f4d91e009cdda08918d114aec5b01220a08010c949c2d1482ce3264dba50cef31102b555366979c49c4fcc9402a0dca533308b6e2659669d66495353c9d41962152642fad9cceae3c7a6b46e580694333946f91ad6d5202e1238d0bf054164764875442cb033c43e67a259784c625f223505401258a4941fa0810321395f2ba92dec060814db62744e2d43809685b41f41e835e56e8b26d53be31957213373f294c34bde0e10a6c3ef42fdd5b479490667acaab0c29b21179a2615849ad03a71c45852424b7003be0b25c43a8fdb01636830ce95110f487dacb7510890a28290f0d33adb4084a272644c0a083030f42f74ea90345cc80f20b515e74924dc1069a003e3193e46912310583cb5e743e1a961ac4228e7521102505b05e8ec68405c6ed57142def56f6c1a25fc10a094cd03643e03a2ddf66a23c4d2cb88bf01c2e82027c2e244fc1d1a2a4bc0e510e66af1a77458a0ca8167a4042e5e4b14dccff436386a70298664edc000428007908828002237c4ba0a13df48ef275289600795a41a558e73f2aa5291e83c17a0416cf4f8c99870016163c622bd941a599568626b36436000423fe84026e2546d46174163961104c169eb450bf3061807feaea25493a1222a27cd13b68c6e20516c0f095fc37d1a1ac94aaf747f03f64c1ad06c1f10a3ac908b4548d50d1f25833c52b3a9c2a261cba8735c1721e0b5ff035338a367a9343106bf1f30116831d2a1a06241062b05e4c867b56b679a29872c50ce45629f68072aec6d88a1660298a762833d67af22cc00d702404bcc9a7f05195f4085aaab9209c6972ec1ca1a1acb344fc564c0650e6f742a50f4401b219f57652deae56dc2e01100a99e028648a079a9c141dc66607d4b832b66d46f507ae721f6379834b72f8300f0c206075d57270316592631c4ff4b21e60c06dc68d4514628b28e377c31ea155b2a6217a200c31d49514bc5440da5ccc19500317266800a84f4fab526da0c0720ef94b2faca2a4157ee1ba65b7e2b151600a881449598d6029198fd496db689262d3c369412116d44a3973c467a4564fd5c616368264ebb56514a2d6872e00f2bcbba7f03287d946c614d7765e5473f33b2fb192b70b37083adc1b99fc185988d5b45c45c145f5025585c71bac349c4c4c34210485350052a5951b630cd84c2ae24ed58f045ac9c349dd350584a4677fd2c058c1da227132d8497441a7ea159fcec89b00298904be41403d94927ab72f2d77030b08f1bb06d428259a6dc06eb361050d65957643b1dc63e96ccd168aa57a86255c9e03ef03dbc6f2f2e0074d2451f0bc4fc2b500c2559d62b514515f030a6c494601519ed2162a631133cb5ea60c1ac6c736461b63480e09c416e95ed45b62cf586cb071223c5f288567a40f636514fdc10e08aabed9ea9c0125ea2f748312b3cc440e1aea4a5c7c24b6fa04b0104b7c4e101c23a24116e7d5423b4f0154e877c28ed536fdd0a1e81eb357624eaf2b5817312e0568edc61ad00e44526d0604ea2c34d60f2f139ec7dd33b252e8319659cdd96ec73de76a8e5b4146100010531c664dc611f12a24977660364584b123530298944965e417c5a365500c1d8b4c3f4fedb5b564b019d31ad573a6843700c36a6daff45ea302ed9f4a6c23a4d448165833324ae08bec6b4b35e880af37ce71542e8bc1b8869b120dc4c4efc1c93cbb997018426dc361a946ad50dd2a6a11baf742cf73950182bfa8e46f1ab1c38f7c94ac299d84199c4882879993a92201043d26078907026a3ba4903c00904b265bd5b938f3a2bf5e00106bc5502206c05203c05b4000030001209025b8c015bf0c3b00700c81807bbc3b85463bb305506bba78020070670140570180048ac19fbf704c83a00605ff76002052632032007a30cd907f06036b65b06db8e0591dec2104a4000725f1880608054250027d2f178600427f302802807880807240c80c84355603903e00ae3286005903860c85f8294527d118c80d8204000660c83048007d6d00006900035000361007aa800751000d87901dc250517c604978000d0004e1e069d69d4b4280470027b2033bc47b2025af9b90e5c03693183694b07731e7a884c06c7ca04501504a409043854eb640c96d01007884a451780056094c0fd6247d7400334a4057a8f80feb020fbe0748b4a25e85470fe4fca055052f0c43103f7c38567a800e956c0c04a408880016094fb769d601751c2790393b0a792fb0c87a2661107c7980227927b27cb7bcfd2f53605ee2fcd052ff2fcaf980204088baffd806f85830066fd1d5b036fcd043360330038fba7b0c4a46af8000652f7cf8e0190c2fa9fff7f1fb0cc9fca406a8948dff221bfd9ff6fd604efe0e1efbefc17f278e063fca07b731facc11c03fda7e4ff4bfab01afed3f5afa1c1ebe8df25faf7c0983d84701080d7e9c052fabfcf1e84f6d02194200b10794be2907e3001ff8594422fff4a03dfcc0072f43d951844037f5cf80034d2e440cfa2f4bb6d40e9fbefd0fedff52f258109ab601205380afe4800afa3fcefe800adfbc7d77e300088bb03340d00c5fb2fcd08abf094a8004903c0defa53113cc80c2063dd9eef125e04a34c20451427a5d1fd65300c0320328013f01054fc67e1294500efd180e1f2906700c408e028025211c194067a1f0169e6e0b9fbe8017e4df5ef8b7cec0660b5c328309e6a0a4051fd908d6f247aca4ac03102e00d001e868f3ad163ca801a834036a1a1008f32010000)  used by other package, react PDF-renderer. They too use this react-pdf as the pdf viewer. 

So the issue is whenever user changes some text on the text editor (left side), the PDF preview on the right side disappears for a split second and shows 'Loading PDF..' message. I understand why this happening but it is very annoying for the user to see the PDF flickering each and every time they make some changes on the text editor. 

So, when the user makes some change on the text editor, is there any way we can 'continue' showing the previously renderered PDF instead of showing the 'Loading PDF..' message while the new PDF is being rendered? 

More info:

Here is how I managed to implement a PDF viewer with pagination using react-pdf, which will display a pdf made using react-pdf/renderer. 

```js
import { pdfjs, Document, Page, Outline } from 'react-pdf';
import { pdf } from '@react-pdf/renderer';

pdfjs.GlobalWorkerOptions.workerSrc = `https://cdnjs.cloudflare.com/ajax/libs/pdf.js/${pdfjs.version}/pdf.worker.js`;

const PDFViewer = ({ children }) => {
  const [pdfUrl, setPdfUrl] = useState(null);
  const [numPages, setNumPages] = useState(null);
  const [pageNumber, setPageNumber] = useState(null);

  const onDocumentLoadSuccess = useCallback((document) => {
    const { numPages: nextNumPages } = document;
    setNumPages(nextNumPages);
    setPageNumber(1);
  }, []);

  const onItemClick = useCallback(
    ({ pageNumber: nextPageNumber }) => setPageNumber(nextPageNumber),
    [],
  );

  const changePage = useCallback(
    (offset) => setPageNumber((prevPageNumber) => (prevPageNumber || 1) + offset),
    [],
  );

  const previousPage = useCallback(() => changePage(-1), [changePage]);

  const nextPage = useCallback(() => changePage(1), [changePage]);

  useEffect(() => {
    const child = React.Children.only(children);

    pdf(child).toBlob().then(blob => {
      setPdfUrl(URL.createObjectURL(blob));
    })
  }, [children]);

  return (
    <Document
      file={pdfUrl}
      onLoadSuccess={onDocumentLoadSuccess}
      onItemClick={onItemClick}
    >
      <Outline
        className="custom-classname-outline"
        onItemClick={onItemClick}
      />
      <Page renderMode='svg' pageNumber={pageNumber} />
      <div className="Test__container__content__controls">
        <button
          disabled={pageNumber <= 1}
          onClick={previousPage}
          type="button"
        >
          Previous
        </button>
        <span>
          {`Page ${pageNumber || (numPages ? 1 : '--')} of ${numPages || '--'}`}
        </span>
        <button
          disabled={pageNumber >= numPages}
          onClick={nextPage}
          type="button"
        >
          Next
        </button>
      </div>
    </Document>
  );
}

export default PDFViewer;
```

### Steps to reproduce

Discussed above

### Expected behavior

Discussed above

### Actual behavior

Discussed above

### Environment

- **Browser (if applicable)**:
- **React-PDF version**:
- **React version**:
- **Webpack version (if applicable)**:

**Comment by wojtekmaj** [comment, marked as answer]
Hey,
A very similar problem already has a solution: show previous `<Page />` when the next one is being rendered. I believe the solution you're looking for will be largely the same, only applied to `<Document />`.

https://codesandbox.io/s/react-pdf-prevent-flash-se4r7s

Since you didn't provide CodeSandbox so I had no working code to work on, I went ahead and implemented this in react-pdf/renderer's REPL itself here: https://github.com/diegomura/react-pdf-site/pull/91

**Comment by hiukky** [comment]
@wojtekmaj  I spent today trying to get a similar result by analyzing the react-pdf implementation but without success. Is there any way of exercising greater control over the rendering of the Document? I have a scenario where I need to load pages of a document in parts in small batches due to data modeling, and in large volume.

I'm using react-pdf for visualization, react-virtuoso for virtualization and pdf-lib for manipulation.

Each document already has metadata with information on the number of pages the document has, so I enter the total number of pages and react-virtuoso calculates the scroll.

Then I create a document (ArrayBuffer ) with the blank pages and after each query (ArrayBuffer[]) I merge the new pages with the previously loaded ones and generate a new Blob to pass to the Document.

This logic is working very well, but the Document always renders again due to this change, displaying the message Loading, or document not found.

I've tried keeping a cache of the rendered document, or some form of memoization, but with these types of immutable data it's not possible.

Is there any way to update the file in this stream scenario without re-rendering the document, or even if it does re-render don't make changes to the UI such as messages etc, keeping what was already rendered before the next render?

I'm trying not to involve Document on every page, even virtualizing will have a cost.

**Reply by nicholsss** [reply]
Were you able to find solutation for this?

**Reply by asadshah777** [reply]
> Were you able to find solutation for this?

I have implemented a custom version of their react-pdf-repl site's CustomPDFRenderer and it wont show that annoying messages, and will perform opacity transitions and have debouncing as well

**Comment by jukkaleh-atoz** [comment]
Had success using loading attribute. This prevented double page rending problem that codesandbox above had.
https://codesandbox.io/s/react-pdf-prevent-flash-se4r7s 

```
export default function PdfDocument(props) {
  const [renderedPageNumber, setRenderedPageNumber] = useState<number | undefined>(undefined);

  return (
    <Document {...props.document} >
      <Page
        key={props.page.pageNumber}
        {...(renderedPageNumber && {loading: <Page key={renderedPageNumber} pageNumber={renderedPageNumber}/>})}
        onRenderSuccess={() => {
          setRenderedPageNumber(props.page.pageNumber);
        }}
        {...props.page}
      />
    </Document>
  );
}
````
